cmd/pod-scaler/producer.go

```diff
@@ -0,0 +1,494 @@
+						errors:  make(chan error),
+						// there's also no chance that Prometheus will be able to handle any real concurrent
+						// request volume, so don't even bother trying to request more samples at once than
+						// a third of the maximum samples it can technically provide in one request
+						sync: semaphore.NewWeighted(MaxSamplesPerRequest / 3),
+						wg:   &sync.WaitGroup{},
+					}
+					wg.Add(1)
+					go func() {
+						defer wg.Done()
+						if err := q.execute(ctx, metadata, now); err != nil {
+							metadata.logger.WithError(err).Error("Failed to query Prometheus.")
+						}
+						metadata.logger.Info("Finished querying Prometheus.")
+					}()
+				}
+				go func() { // don't associate this with the context as we want to flush when interrupted
+					wg.Wait()
+					if err := StoreCache(dataCache, name, cache, logger); err != nil {
+						logger.WithError(err).Error("Failed to write cached data.")
+					}
+				}()
+			}
+
+			select {
+			case <-ctx.Done():
+				logrus.Debug("Interrupted, cancelling.")
+				return
+			case <-ticker.C:
+			}
+		}
+	})
+}
+
+// queryFor applies our filtering and left joins to a metric to get data we can use
+func queryFor(metric string) string {
+	return `sum by (namespace,pod,container) (` + metric + `) * on(namespace,pod) 
+  group_left(
+    ` + string(LabelNameOrg) + `,
+    ` + string(LabelNameRepo) + `,
+    ` + string(LabelNameBranch) + `,
+    ` + string(LabelNameVariant) + `,
+    ` + string(LabelNameTarget) + `,
+    ` + string(LabelNameStep) + `
+  ) max by (
+    namespace,
+    pod,
+    ` + string(LabelNameOrg) + `,
+    ` + string(LabelNameRepo) + `,
+    ` + string(LabelNameBranch) + `,
+    ` + string(LabelNameVariant) + `,
+    ` + string(LabelNameTarget) + `,
+    ` + string(LabelNameStep) + `
+  ) (kube_pod_labels{` + string(LabelNameOrg) + `!="",label_created_by_ci="true"})`
+}
+
+func rangeFrom(r prometheusapi.Range) TimeRange {
+	return TimeRange{
+		Start: r.Start,
+		End:   r.End,
+	}
+}
+
+type querier struct {
+	lock *sync.RWMutex
+	data *CachedQuery
+}
+
+type clusterMetadata struct {
+	logger *logrus.Entry
+	name   string
+	client prometheusapi.API
+	errors chan error
+
+	lock    *sync.RWMutex
+	maxSize int64
+
+	// sync guards the number of concurrent samples we can be asking Prometheus for at any one time
+	sync *semaphore.Weighted
+	wg   *sync.WaitGroup
+}
+
+func (q *querier) execute(ctx context.Context, c *clusterMetadata, until time.Time) error {
+	runtime, err := c.client.Runtimeinfo(ctx)
+	if err != nil {
+		return fmt.Errorf("could not query Prometheus runtime info: %w", err)
+	}
+	retention, err := model.ParseDuration(runtime.StorageRetention)
+	if err != nil {
+		return fmt.Errorf("could not determine Prometheus retention duration: %w", err)
+	}
+	r := prometheusapi.Range{
+		Start: until.Add(-time.Duration(retention)),
+		End:   until,
+		Step:  1 * time.Minute,
+	}
+
+	errLock := &sync.Mutex{}
+	var errs []error
+	go func() {
+		errLock.Lock()
```

Comment:
Nit: Not using `defer` can easily lead to hard to debug locking issues down the line if someone changes the content of the closure